server/coordinator.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+const (
+	regionCacheTTL     = 2 * time.Minute
+	historiesCacheSize = 1000
+	eventsCacheSize    = 1000
+)
+
+type coordinator struct {
+	sync.RWMutex
+
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	cluster *clusterInfo
+	opt     *scheduleOption
+	checker *replicaChecker
+
+	operators map[ResourceKind]map[uint64]*balanceOperator
+
+	regionCache *expireRegionCache
+	histories   *lruCache
+	events      *fifoCache
+}
+
+func newCoordinator(cluster *clusterInfo, opt *scheduleOption) *coordinator {
+	c := &coordinator{
+		cluster:     cluster,
+		opt:         opt,
+		checker:     newReplicaChecker(cluster, opt),
+		regionCache: newExpireRegionCache(regionCacheTTL, regionCacheTTL),
+		histories:   newLRUCache(historiesCacheSize),
+		events:      newFifoCache(eventsCacheSize),
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.TODO())
+
+	c.operators = make(map[ResourceKind]map[uint64]*balanceOperator)
+	c.operators[leaderKind] = make(map[uint64]*balanceOperator)
+	c.operators[storageKind] = make(map[uint64]*balanceOperator)
+
+	return c
+}
+
+func (c *coordinator) dispatch(region *regionInfo) *pdpb.RegionHeartbeatResponse {
+	op := c.getOperator(region.GetId())
+	if op == nil {
+		op = c.checker.Check(region)
+	}
+	if op == nil {
+		return nil
+	}
+
+	ctx := newOpContext(c.hookStartEvent, c.hookEndEvent)
+	finished, res, err := op.Do(ctx, region)
+	if err != nil {
+		log.Errorf("failed to do operator for region %v: %v", region, err)
+	}
+	if err != nil || finished {
+		c.removeOperator(op)
+	}
+	return res
+}
+
+func (c *coordinator) run() {
+	c.wg.Add(1)
+	go c.runScheduler(newLeaderBalancer(c.opt), newLeaderController(c))
+
+	c.wg.Add(1)
+	go c.runScheduler(newStorageBalancer(c.opt), newStorageController(c))
+}
+
+func (c *coordinator) stop() {
+	c.cancel()
+	c.wg.Wait()
+}
+
+func (c *coordinator) runScheduler(s Scheduler, ctrl Controller) {
+	defer c.wg.Done()
```

Comment:
We should call `wg.Add(1)` before creating a goroutine to avoid data race, because the goroutine may run after we call `wg.Wait()`.